nautobot/extras/models/secrets.py

```diff
@@ -0,0 +1,170 @@
+import logging
+
+from django.core.exceptions import ValidationError
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db import models
+from django.urls import reverse
+
+from jinja2 import Environment, StrictUndefined
+from jinja2.exceptions import UndefinedError, TemplateSyntaxError
+
+from nautobot.core.fields import AutoSlugField
+from nautobot.core.models import BaseModel
+from nautobot.core.models.generics import OrganizationalModel, PrimaryModel
+from nautobot.extras.choices import SecretsGroupAccessTypeChoices, SecretsGroupSecretTypeChoices
+from nautobot.extras.registry import registry
+from nautobot.extras.secrets.exceptions import SecretError, SecretParametersError, SecretProviderError
+from nautobot.extras.utils import extras_features
+
+
+logger = logging.getLogger(__name__)
+
+
+@extras_features(
+    "custom_fields",
+    "custom_links",
+    "custom_validators",
+    "graphql",
+    "relationships",
+    "webhooks",
+)
+class Secret(PrimaryModel):
+    """
+    Data model providing access to a "secret" such as a device credential or a systems-integration token.
+
+    Note that this model **does not** STORE the actual secret data, rather it provides ACCESS to the secret for other
+    Nautobot models and APIs to make use of as needed and appropriate.
+    """
+
+    name = models.CharField(max_length=100, unique=True)
+    slug = AutoSlugField(populate_from="name")
+    description = models.CharField(max_length=200, blank=True)
+    provider = models.CharField(max_length=100)
+    parameters = models.JSONField(encoder=DjangoJSONEncoder, default=dict)
+
+    csv_headers = [
+        "name",
+        "slug",
+        "description",
+        "provider",
+        "parameters",
+    ]
+    clone_fields = [
+        "provider",
+    ]
+
+    class Meta:
+        ordering = ["name"]
+
+    def __str__(self):
+        return self.name
+
+    def get_absolute_url(self):
+        return reverse("extras:secret", args=[self.slug])
+
+    def to_csv(self):
+        return (
+            self.name,
+            self.slug,
+            self.description,
+            self.provider,
+            self.parameters,
+        )
+
+    def rendered_parameters(self, obj=None):
+        """Render self.parameters as a Jinja2 template with the given object as context."""
+        environment = Environment(undefined=StrictUndefined)
```

Comment:
IIRC the reason for doing this this way was to be able to set `undefined=StrictUndefined` so that undefined references cause the render to fail instead of rendering the undefined references as empty strings.

I suppose I could add a new "jinja_strict" template engine in settings.py and use that if you think it'd be preferable?

Reply:
I think it would be nice to have the consistency in all places we allow jinja templating, so there are no surprises.